src/mm-file-upload/mm-file-upload.js

```diff
@@ -0,0 +1,152 @@
+(function(scope) {
+
+	scope.FileUpload = Polymer({
+		is: 'mm-file-upload',
+
+		properties: {
+			accept: {
+				type: String,
+				value: ''
+			},
+
+			useDataUrls: {
+				type: Boolean,
+				value: false
+			},
+
+			multiple: {
+				type: Boolean,
+				value: false
+			},
+
+			disabled: {
+				type: String,
+				value: false,
+				reflectToAttribute: true
+			},
+
+			dragDropEnabled: {
+				type: Boolean,
+				value: false
+			},
+
+			error: {
+				type: Boolean,
+				value: false
+			},
+			
+			files: {
+				type: Array,
+				value: function() { return []; },
+				notify: true
+			},
+
+			fileNames: {
+				type: String,
+				computed: '_getFileNames(files.splices)',
+				readOnly: true
+			},
+
+			placeholder: {
+				type: String,
+				value: 'No file chosen'
+			},
+
+			browseLabel: {
+				type: String,
+				value: 'Browse'
+			}
+		},
+
+		listeners: {
+			'dragenter': '_handleDrag',
+			'dragover': '_handleDrag',
+			'drop': '_handleDrop'
+		},
+
+		_getFileNames: function(changeRecord) {
+			return this.files.length > 0 ? this.files.map(function(f) { return f.name; }).join(', ') : '';
+		},
+
+		_handleFiles: function(e) {
+			if(e.target.files.length > 0) {
+				var files = Array.prototype.slice.call(e.target.files),
+					filtered = files.filter(this._validateFile.bind(this));
+
+				if(files.length !== filtered.length) this.error = true;
+				else this.error = false;
+
+				if(this.useDataUrls) {
+					for(var i=filtered.length-1; i>=0; i--) {
+						(function(file) {
+
+							var reader = new FileReader();
+							reader.onload = function() {
+								this.push('dataUrls',reader.result);
```

Comment:
When adding a .png to the "Using base64 encoding" example - Polymer chokes on push... looks like dataUrls needs a default of [] (?)